tools/plugin/alsaplug/tplg_ctl.c

```diff
@@ -0,0 +1,204 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2024 Intel Corporation. All rights reserved.
+//
+
+/* SOF topology kcontrols */
+
+#include <stdio.h>
+#include <sys/poll.h>
+#include <string.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <sys/wait.h>
+#include <sys/stat.h>
+#include <mqueue.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <semaphore.h>
+#include <assert.h>
+#include <errno.h>
+#include <math.h>
+
+#include <alsa/asoundlib.h>
+#include <alsa/pcm_external.h>
+
+#include <rtos/sof.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/component.h>
+#include <ipc/stream.h>
+#include <tplg_parser/topology.h>
+
+#include "plugin.h"
+#include "common.h"
+
+#define SOF_IPC4_VOL_ZERO_DB	0x7fffffff
+#define VOLUME_FWL      16
+/*
+ * Constants used in the computation of linear volume gain
+ * from dB gain 20th root of 10 in Q1.16 fixed-point notation
+ */
+#define VOL_TWENTIETH_ROOT_OF_TEN	73533
+/* 40th root of 10 in Q1.16 fixed-point notation*/
+#define VOL_FORTIETH_ROOT_OF_TEN	69419
+
+/* 0.5 dB step value in topology TLV */
+#define VOL_HALF_DB_STEP	50
+
+/*
+ * Function to truncate an unsigned 64-bit number
+ * by x bits and return 32-bit unsigned number. This
+ * function also takes care of rounding while truncating
+ */
+static uint32_t vol_shift_64(uint64_t i, uint32_t x)
+{
+	/* do not truncate more than 32 bits */
+	if (x > 32)
+		x = 32;
+
+	if (x == 0)
+		return (uint32_t)i;
```

Comment:
nitpick: I'd put this test first - then you don't need to check `> 32` if `x == 0`